### PR TITLE
[dagster-airlift][sensor-pluggability][rfc] revamped sensor pluggability

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary & Motivation
Another initial PR to level set on top-level sensor API. This attempt opts for a more simple top level API, which allows for transformation of the constructed materializations, and that's about it. 
Open questions:
- An interesting implication here I think is; in this world, where do partitions slot in? Should they slot in in the "default" event transformer?
- How should the user interact with the airflow mappings? The `AirflowDefinitionsData` object at this point can be user-constructed pretty easily and used throughout an implementation of an event transformer, but does that feel janky?
- Is it the right behavior that we allow the user to change the ordering?

## How I Tested These Changes
Added a basic unit test for pluggability
## Changelog
NOCHANGELOG

### CODE REVIEW

**Comment by dpeng817** [comment]
* **#25015** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25015?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/sensor_builder.py [hunk 1]
```diff
@@ -125,10 +125,17 @@ def airflow_dag_sensor(context: SensorEvaluationContext) -> SensorResult:
...
+            event_transformation_fn(all_events) if event_transformation_fn else all_events
+        )
+
         context.update_cursor(serialize_value(new_cursor))
+        sorted_events = sorted_asset_events(transformed_events, repository_def)
 
         return SensorResult(
-            asset_events=sorted_asset_events(all_materializations, repository_def),
+            asset_events=event_transformation_fn(sorted_events)
+            if event_transformation_fn
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/sensor/sensor_builder.py:137]: did you mean to call event_transformation_fn twice?

Also just in terms of readability and navigability I think we should directly call the "default" function here so you can jump right to it in the IDE?

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/sensor_builder.py [hunk 2]
```diff
@@ -138,23 +145,24 @@ def airflow_dag_sensor(context: SensorEvaluationContext) -> SensorResult:
...
     repository_def: RepositoryDefinition,
 ) -> List[AssetMaterialization]:
     """Sort materializations by end date and toposort order."""
     topo_aks = repository_def.asset_graph.toposorted_asset_keys
+    timestamps_and_events = [(get_timestamp_from_asset_event(event), event) for event in events]
     return [
         sorted_mat[1]
         for sorted_mat in sorted(
-            all_materializations, key=lambda x: (x[0], topo_aks.index(x[1].asset_key))
+            timestamps_and_events, key=lambda x: (x[0], topo_aks.index(x[1].asset_key))
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/sensor/sensor_builder.py:157]: a comment that indicates what this sorted order ends up being would be good. lambda is hard to parse

**Review by schrockn — COMMENTED** [review]
I think this approach overall makes sense 👍🏻
